nixos/modules/tasks/auto-upgrade.nix

```diff
@@ -174,6 +253,251 @@ in
         '';
       };
 
+      desync = lib.mkOption {
+        default = { };
+        description = ''
+          A set of nixpkgs copies that can be updated independently of the main nixpkgs.
+
+          It allows the system to keep building even if the latest version of the channel
+          has a broken build for a package.
+
+          If it is used, then the `nixos-upgrade-desyncs` command is added to the system
+          environment, that can be used to try to upgrade the desyncs to the current channel
+          version.
+          If they build fine then they will be upgraded, if not they will be left at their
+          current setting.
+
+          Using `maxDesyncAge`, you can limit the number of rebuilds that one nixpkgs
+          version is allowed to span.
+          For example, if you have `system.autoUpgrade` set to upgrade once a day, and
+          `maxDesyncAge` set to `14` (the default), then your packages will always be on
+          a version of `nixpkgs` that is at most two weeks old.
+          If no such version has a passing build for one week, then the `mixos-rebuild`
+          service will still fail.
+
+          Finally, note that overlays are not taken into account inside desyncs. This is
+          because, otherwise, evaluation would enter infinite loops.
+          If you need to overlay your desyncs, you can use the `desync.*.overlays` option.
```

Comment:
Is there a reason we can't just read the `nixpkgs.overlays` option instead of introducing a new `desync.*.overlays` option?

If the new option stays, this reference to it should ideally be a markdown link. Also, this again feels like maybe it should be an admonition.

Reply:
Oh, I see. You're using the `nixpkgs.overlays` to implement your desynced packages.

Maybe we could avoid this inf-rec by having an internal option: `nixpkgs.finalOverlays` or something?

The `nixpkgs` module would simply define `nixpkgs.finalOverlays = cfg.overlays`, while other internal implementation, like the desync module, could additionally write to the `nixpkgs.finalOverlays` option, while still reading from the `nixpkgs.overlays` option.

Even so, it feels like some users would want their overlays to be considered in desyncs while other users may not. So we'd probably also need some boolean toggle.

I guess this idea is fine to be left out of scope for this initial PR. If you like the idea, maybe add a TODO comment?